### PR TITLE
NIST SP800-56A co-factor ECDH KATs

### PR DESCRIPTION
**Pull request by bbbrumley**
Adds NIST SP800-56A co-factor ECDH test vectors. KATs taken from

http://csrc.nist.gov/groups/STM/cavp/component-testing.html#ecc-cdh

http://csrc.nist.gov/groups/STM/cavp/documents/components/ecccdhtestvectors.zip

### CODE REVIEW

**Comment by richsalz** [comment]
+1

**Comment by kroeckx** [comment]
+1

File: test/ecdhtest.c [hunk 1]
```diff
@@ -462,6 +461,105 @@ static int ecdh_kat(BIO *out, const ecdh_kat_t *kat)
...
+    /* (QIUTx, QIUTy) is IUT's public key */
+    if(!BN_hex2bn(&x, kat->QIUTx))
+        goto err;
+    if(!BN_hex2bn(&y, kat->QIUTy))
+        goto err;
+    if (is_char_two) {
+#ifndef OPENSSL_NO_EC2M
+        if (!EC_POINT_set_affine_coordinates_GF2m(group, pub, x, y, NULL))
+#endif
+            goto err;
```
Comment (kroeckx) [inline test/ecdhtest.c]: Should we skip the test instead of generating an error in this case?

Comment (richsalz) [inline test/ecdhtest.c]: Yes, we probably should.  And that avoids breaking a statement with ifdef :)

Comment (bbbrumley) [inline test/ecdhtest.c]: Thanks for looking.

Hmm I don't see any easy way to test if it's a binary curve from just the NID. EC_GROUP_new_by_curve_name will fail before I can even fetch the field type.

Currently, this is handled in ecdhtest_cavs.h with ifdef statements, so the binary tests only get compiled in if binary curves are enabled.

Is this sufficient or do you want something more elaborate? BTW tested with both no-ec2m on and off.

Comment (richsalz) [inline test/ecdhtest.c]: Okay, as long as you tested it both ways and there are no errors, that’s good.  But can you not break the if/goto and instead duplicate the goto?

**Comment by kroeckx** [comment]
That's not exactly what I had in mind.  More like:

    #ifdef OPENSSL_NO_EC2M
        if (is_char_two) {
            rv = -1
            goto err;
        }
    #endif

And maybe print skip in that case.

File: test/ecdhtest.c [hunk 2]
```diff
@@ -462,6 +461,105 @@ static int ecdh_kat(BIO *out, const ecdh_kat_t *kat)
...
+    EC_KEY_free(key1);
+    EC_POINT_free(pub);
+    BN_free(bnz);
+    BN_free(x);
+    BN_free(y);
+    OPENSSL_free(Ztmp);
+    OPENSSL_free(Z);
+    if (rv)
+        BIO_puts(out, " ok\n");
+    else {
```
Comment (kroeckx) [inline test/ecdhtest.c]: Please use { } for the if when the else has them.

File: test/ecdhtest.c [hunk 3]
```diff
@@ -462,6 +461,105 @@ static int ecdh_kat(BIO *out, const ecdh_kat_t *kat)
     return rv;
 }
 
+#include "ecdhtest_cavs.h"
+
+static int ecdh_cavs_kat(BIO *out, const ecdh_cavs_kat_t *kat)
```
Comment (kroeckx) [inline test/ecdhtest.c]: Can you add some small comment about what this function tests and what the return value is?

File: test/ecdhtest.c (hunk 1 above)
Comment (kroeckx) [inline test/ecdhtest.c]: On Mon, Feb 29, 2016 at 11:07:18AM -0800, Billy Brumley wrote:
> 
> Hmm I don't see any easy way to test if it's a binary curve from just the NID. EC_GROUP_new_by_curve_name will fail before I can even fetch the field type.
> 
> Currently, this is handled in ecdhtest_cavs.h with ifdef statements, so the binary tests only get compiled in if binary curves are enabled.
> 
> Is this sufficient or do you want something more elaborate? BTW tested with both no-ec2m on and off.

For some reason this comment doesn't seem to be showing up, but if
you just don't test them with ifdefs in the .h file that's fine
for me.

**Comment by bbbrumley** [comment]
@kroeckx It won't actually reach your is_char_two branch because EC_GROUP will be NULL. Right now it's avoided by ifdef blocks in ecdhtest_cavs.h separating the test cases. Can you recommend a smoother way, considering all we have going in is the NID?

**Comment by kroeckx** [comment]
So I already mailed I was fine with the test being excluded with #ifdef's in the .h file.

For some reason I can't find the other comments about this.

Anyway, if you could just change those other small things.

**Comment by bbbrumley** [comment]
@kroeckx @richsalz Thanks for the comments. Let me know if there's anything else.

**Comment by kroeckx** [comment]
:+1:

**Comment by richsalz** [comment]
pushed, thanks!

**Pull request closed without merging**